marshaller.go

```diff
@@ -0,0 +1,64 @@
+package commando
+
+import (
+	"encoding/csv"
+	"reflect"
+)
+
+// Marshaller is a CSV to struct marshaller.
+type Marshaller struct {
+	config *validConfig
+	writer *csv.Writer
+}
+
+// NewMarshaller is a convenience function which allocates and
+// returns a new Marshaller.
+func NewMarshaller(holder interface{}, writer *csv.Writer) (*Marshaller, error) {
+	return (&Config{Holder: holder}).NewMarshaller(writer)
+}
+
+// NewMarshaller creates a marshaller from a csv.Writer.
+func (c *Config) NewMarshaller(writer *csv.Writer) (*Marshaller, error) {
+	vc, err := c.validate(nil)
+	if err != nil {
+		return nil, err
+	}
+
+	m := &Marshaller{
+		writer: writer,
+		config: vc,
+	}
+
+	if err := m.writeHeaders(); err != nil {
+		return nil, err
+	}
+	return m, nil
+}
+
+func (m *Marshaller) writeHeaders() error {
+	return m.writer.Write(m.config.structInfo.headers())
+}
+
+func (m *Marshaller) Write(record interface{}) error {
```

Comment:
just a general thought while I see all these `interface{}`s pop up. Would it make more sense to have typed interfaces that were required to implement a Write() or similar function instead of passing blanket types?

Reply:
maybe something that instead of actually /writing/ it produced a normalized structure that this function could process. 
hear me out and I am aware I am possibly misunderstanding the intent, as well as simplifying a bit. (typing this directly in the edit box here, so be kind)

```go

type Row map[string]string; // probably something more effective to preserve ordering like an idxhash or simply this:

type Row struct {
    Columns []string
    Elements []string
}

type Record interface {
    Yield() (Row, error)
}

func (m *Marshaller) Write(record Record) {
    row := record.Yield();
    // most of what's left of your function here
}
```

Reply:
This is pretty close to what `encoding/csv` does already, and I believe there's another library that turns things into `map[string]string`.  We have a ton of code that relies on gocsv using reflection to write structs out as CSV, so adopting something like this would mean writing a bunch of manual conversion functions.

In short, I think that's a reasonable design for a generic library started from scratch, but my aim is to take the stuff in gocsv that we rely on and make it less annoying to use, which I think that'd compromise.

Reply:
ack!